packages/rich-text-structure/src/create.js

```diff
@@ -0,0 +1,326 @@
+/**
+ * Internal dependencies
+ */
+
+import { isEmpty } from './is-empty';
+
+/**
+ * Browser dependencies
+ */
+
+const { TEXT_NODE, ELEMENT_NODE } = window.Node;
+
+/**
+ * Parse the given HTML into a body element.
+ *
+ * @param {string} html The HTML to parse.
+ *
+ * @return {HTMLBodyElement} Body element with parsed HTML.
+ */
+function createElement( html ) {
+	const htmlDocument = document.implementation.createHTMLDocument( '' );
+
+	htmlDocument.body.innerHTML = html;
+
+	return htmlDocument.body;
+}
+
+/**
+ * Creates rich text value and selection objects from a DOM element and range.
+ *
+ * @param {HTMLElement} element      Element to create value object from.
+ * @param {Range}       range        Range to create selection object from.
+ * @param {string}      multilineTag Multiline tag if the structure is multiline.
+ * @param {Object}      settings     Settings passed to `createRecord`.
+ *
+ * @return {Object} A rich text record.
+ */
+export function create( element, range, multilineTag, settings ) {
+	if ( typeof element === 'string' ) {
+		element = createElement( element );
+	}
+
+	if ( ! multilineTag ) {
+		return createRecord( element, range, settings );
+	}
+
+	const emptyRecord = {
+		formats: [],
+		text: '',
+	};
+
+	if ( ! element || ! element.hasChildNodes() ) {
+		return emptyRecord;
+	}
+
+	const children = Array.from( element.children )
+		.filter( ( { nodeName } ) => nodeName.toLowerCase() === multilineTag );
+	const maxIndex = children.length - 1;
+
+	if ( maxIndex < 0 ) {
+		return emptyRecord;
+	}
+
+	return children.reduce( ( accumulator, node, index ) => {
+		const { start, end, text, formats } = createRecord( node, range, settings );
+		const length = accumulator.text.length;
+
+		if ( range ) {
+			if ( start !== undefined ) {
+				accumulator.start = length + start;
+			} else if (
+				node.parentNode === range.startContainer &&
+				node === range.startContainer.childNodes[ range.startOffset ]
+			) {
+				accumulator.start = length;
+			}
+
+			if ( end !== undefined ) {
+				accumulator.end = length + end;
+			} else if (
+				node.parentNode === range.endContainer &&
+				node === range.endContainer.childNodes[ range.endOffset - 1 ]
+			) {
+				accumulator.end = length + text.length;
+			}
+		}
+
+		accumulator.formats = accumulator.formats.concat( formats );
+		accumulator.text += text;
+
+		if ( index !== maxIndex ) {
+			accumulator.formats = accumulator.formats.concat( [ , , ] );
+			accumulator.text += '\n\n';
+		}
+
+		return accumulator;
+	}, emptyRecord );
+}
+
+/**
+ * Creates a rich text value object from a DOM element.
+ *
+ * @param {HTMLElement} element      Element to create value object from.
+ * @param {string}      multilineTag Multiline tag.
```

Comment:
what is `multilineTag`? the description is recursive

maybe something explanatory like "HTML tag name used to indicate that rich text has multiple lines" or "HTML tag name used to separate multiple lines in rich text, e.g. 'p' or 'br'" (I'm not exactly sure what it is so I can't provide good examples of what I'm trying to suggest)